cli/tools/test/channel.rs

```diff
@@ -0,0 +1,481 @@
+  (factory.worker(), receiver)
+}
+
+/// Polls for the next [`TestEvent`] from any worker. Events from multiple worker
+/// streams may be interleaved.
+pub struct TestEventReceiver {
+  receiver: UnboundedReceiver<(usize, TestEvent)>,
+}
+
+impl TestEventReceiver {
+  /// Receive a single test event, or `None` if no workers are alive.
+  pub async fn recv(&mut self) -> Option<(usize, TestEvent)> {
+    self.receiver.recv().await
+  }
+}
+
+struct TestStream {
+  id: usize,
+  which: TestStdioStream,
+  read_opt: Option<AsyncPipeRead>,
+  sender: UnboundedSender<(usize, TestEvent)>,
+}
+
+impl TestStream {
+  fn new(
+    id: usize,
+    which: TestStdioStream,
+    pipe_reader: PipeRead,
+    sender: UnboundedSender<(usize, TestEvent)>,
+  ) -> std::io::Result<Self> {
+    // This may fail if the tokio runtime is shutting down
+    let read_opt = Some(pipe_reader.into_async()?);
+    Ok(Self {
+      id,
+      which,
+      read_opt,
+      sender,
+    })
+  }
+
+  /// Send a buffer to the test event channel. If the channel no longer exists, shut down the stream
+  /// because we can't do anything.
+  #[must_use = "If this returns false, don't keep reading because we cannot send"]
+  fn send(&mut self, buffer: Vec<u8>) -> bool {
+    if buffer.is_empty() {
+      true
+    } else if self
+      .sender
+      .send((self.id, TestEvent::Output(self.which, buffer)))
+      .is_err()
+    {
+      self.read_opt.take();
+      false
+    } else {
+      true
+    }
+  }
+
+  fn is_alive(&self) -> bool {
+    self.read_opt.is_some()
+  }
+
+  /// Attempt to read from a given stream, pushing all of the data in it into the given
+  /// [`UnboundedSender`] before returning.
+  async fn pipe(&mut self) {
+    let mut buffer = [0_u8; BUFFER_SIZE];
+    let mut buf = ReadBuf::new(&mut buffer);
+    let res = {
+      // No more stream, so just return.
+      let Some(stream) = &mut self.read_opt else {
+        return;
+      };
+      poll_fn(|cx| Pin::new(&mut *stream).poll_read(cx, &mut buf)).await
+    };
+    match res {
+      Ok(_) => {
+        let buf = buf.filled().to_vec();
+        if buf.is_empty() {
+          // The buffer may return empty in EOF conditions and never return an error,
+          // so we need to treat this as EOF
+          self.read_opt.take();
+        } else {
+          // Attempt to send the buffer, marking as not alive if the channel is closed
+          _ = self.send(buf);
+        }
+      }
+      Err(_) => {
+        // Stream errored, so just return and mark this stream as not alive.
+        _ = self.send(buf.filled().to_vec());
+        self.read_opt.take();
+      }
+    }
+  }
+
+  /// Read and "block" until the sync markers have been read.
+  async fn read_until_sync_marker(&mut self) {
+    let Some(file) = &mut self.read_opt else {
+      return;
+    };
+    let mut flush = Vec::with_capacity(BUFFER_SIZE);
+    loop {
+      let mut buffer = [0_u8; BUFFER_SIZE];
```

Comment:
Do we need to create a buffer for each iteration? Maybe we could reuse a buffer?

Reply:
It's a stack buffer -- it's free.